ci: Implement Replay Checker

helmutbuhler opened the pull request:
Fixes #1201
Initial attempt here: #1239

This PR adds automatic checking of replays for compatibility to CI.
It also adds an additional configuration `vc6-releaselog` to CMakePresets.json. It compiles the game with optimizations, logging and crashing enabled. That configuration is still compatible and also checked in the CI.

It also adds two more headless checks:
- One disables a dialog that appears when the music cannot be loaded (not included in the gamedata files CI uses)
- The other is needed because `MessageBoxWrapper` since a recent change no longer checks whether there is a window in the current thread.

The CI internally downloads some gamedata files which are needed to simulate the replay. These gamedata files are trimmed of files which are not needed in headless mode (audio, textures, GUI). They are also encrypted so only the CI runner can access them.

The replays and maps the CI uses to check are also added here to the repository.

The new replaycheck job happens after the VC6 build is complete. Luckily, the VC6 build is currently much faster than the VS22 build and as such the replaycheck and the VS22 build can run in parallel. This PR therefore doesn't increase the runtime of CI much (about a minute).

Before this can be merged, these things still need to happen:
- ~~ArcticDolphin needs to upload the file to s3~~
- ~~Issue #1365 needs to be fixed~~

xezon commented on .github/workflows/check-replays.yml:
```diff
@@ -0,0 +1,206 @@
+              exit 1
+          }
+          Write-Host "Downloading GameData" -ForegroundColor Cyan
+          aws s3 cp s3://github-ci/gen_gamedata_trimmed.7z gen_gamedata_trimmed.7z --endpoint-url $env:AWS_ENDPOINT_URL
```
How big is this data? Will this download anew on every run or is this kept cached on the ci runner?

helmutbuhler replied:
105 MB. It's only downloaded once when the PR will be merged and after that loaded from the cache.

xezon commented on .github/workflows/check-replays.yml:
```diff
@@ -0,0 +1,206 @@
+          EXPECTED_HASH: "3BE86F84B1E771BEA01BD2DBA39442683381AE4401716E9AE9D4AD95C8F45372"
+        shell: pwsh
+        run: |
+          # Download trimmed gamedata of both ZH and Generals. This data cannot be used for playing because it's
```
"Generals 1.08 and Generals Zero Hour 1.04"

I suggest to always name Generals first for consistency.

xezon commented on .github/workflows/check-replays.yml:
```diff
@@ -0,0 +1,206 @@
+              exit 1
+          }
+          Write-Host "Downloading GameData" -ForegroundColor Cyan
+          aws s3 cp s3://github-ci/gen_gamedata_trimmed.7z gen_gamedata_trimmed.7z --endpoint-url $env:AWS_ENDPOINT_URL
```
perhaps name this file

"generals108_zh104_gamedata_trimmed.7z"

to make clear that this file contains not just Generals.

xezon replied:
Also, perhaps it makes sense to split this into 2 files:

"generals108_gamedata_trimmed.7z"
"zerohour104_gamedata_trimmed.7z"

This then gives the freedom to combine this with another zero hour data version. Just a bit more flexible.

helmutbuhler replied:
If @tintinhamans does that, I'll update the code.

xezon replied:
I would do as suggested above. Is flexible and identifiable.

xezon commented on GeneralsMD/Test/Replays/366648.rep:
Does any of these replays contain uncensored IP addresses?

Note that Replay headers contain IP addresses. We probably want to eliminate these before merging them into the repository.

The IP Addresses in GenTool Replays are zeroed in the lower 12 bits.

helmutbuhler replied:
I agree that it's insane...
I haven't checked whether they have valid ip addresses, but all replays (except for the Golden one) are from the gentool server, so they should be "censored".
Do you have a tool to reset the ip address of a replay? If so you can push a commit with new replays. Just make sure the ip is set to something !=0 because there is some code that checks for that. (Maybe set it to 1 or 42).

helmutbuhler replied:
It's binary

helmutbuhler replied:
Maybe you wanna try it? I can give you permission to push to this branch.

JoKeRZH429 replied:
I checked. Golden Replay is not censored.

xezon commented on GeneralsMD/Test/README.md:
```diff
@@ -0,0 +1,14 @@
+
+You can also test with these replays locally:
+- Copy the replays into a subfolder in your `*userpath*/Documents/Command and Conquer Generals Zero Hour Data/Replays` folder.
+- Copy the maps into `*userpath*/Documents/Command and Conquer Generals Zero Hour Data/Maps`
```
`%USERPROFILE%\Documents\...`

xezon commented on Generals/Code/GameEngine/Source/Common/System/Debug.cpp:
```diff
@@ -148,11 +148,14 @@ static void doStackDump();
 #endif
+	if (TheGlobalData && TheGlobalData->m_headless)
+		ignore = true;
+	return ignore;
```
I think this can be further simplified:

```cpp
inline Bool ignoringAsserts()
{
	if (!DX8Wrapper_IsWindowed)
		return true;
	if (TheGlobalData && TheGlobalData->m_headless)
		return true;
#ifdef DEBUG_CRASHING
	if (TheGlobalData && TheGlobalData->m_debugIgnoreAsserts)
		return true;
#endif

	return false;
}

xezon reviewed (commented):
Very nice.

@OmniBlade Are you ok with 3 MB of game user maps and replays in code repo? Or should this be stored a different way?

xezon commented on GeneralsMD/Code/GameEngine/Source/Common/System/Debug.cpp:
For cleaner separation of changes, I think it would be good to have 2 commits go into main branch:

1. The C++ code changes
2. yml edits and binary files additions

helmutbuhler commented:
@tintinhamans Thank you for uploading the datafile to S3 and the help!

@xezon and @Caball009 Thank you for fixing #1365!

I just pulled main and also implemented all suggestions (except for the ip stuff). The CI works now! I also extracted the code changes as suggested into a new PR #1378.

For security reasons, the secrets are not visible for PRs, that's why the CI still seems to fail here. But once it's merged (xezon will have to override the CI check), it should work. After the merge it will also work for all other PRs because the gamedata will be in the cache and not require the secrets anymore.

If you want to check what the CI run will look like, you can see it here: https://github.com/helmutbuhler/CnC_Generals_Zero_Hour/pull/1
(I also tested it there with mismatching builds, timeouts and all those edge cases)

xezon commented on .github/workflows/ci.yml:
```diff
@@ -143,3 +163,19 @@ jobs:
     secrets: inherit
+
+  replaycheck-generalsmd:
+    name: Replay Check MD${{ matrix.preset && '' }}
```
"Replay Check GeneralsMD"

aliendroid1 commented:
Why not just store it in something like Google drive

xezon commented:
It is already bad enough to have things in Amazon cloud. I suggest to not spread data around and keep most things inside the github organization when reasonable. It makes it easier to access and maintain.

I have no access to the Amazon cloud.

xezon commented:
GeneralsGameData is also git. Just in a separate repo. The advantage is that we can just keep adding more and more data there without any limitations.

OmniBlade commented:
If its a another git repo it can potentially be pulled in as a submodule or fetchcontent. You could also potentially abuse the github attachments to create a URL to it that is associated with the repo but not included in the git repo.

xezon commented:
I can help with that. It will be also a good learning excercise because I want to revisit and consider submodules for the Patch repository. (This code repo likely will be a submodule)

helmutbuhler commented:
> @helmutbuhler Check [TheSuperHackers/GeneralsReplays#1](https://github.com/TheSuperHackers/GeneralsReplays/pull/1)

I left a comment there.

xezon commented:
"GeneralsReplays" submodule was added.

helmutbuhler commented:
Ok, I updated it again. The CI now uses the replays and maps here: https://github.com/TheSuperHackers/GeneralsReplays
The gamedata in S3 has also been renamed and split into two archives.

xezon commented:
> One or more required secrets are not set or are empty. R2_ACCESS_KEY_ID: False, R2_SECRET_ACCESS_KEY: False, R2_ENDPOINT_URL: False

@tintinhamans Can you supply these things whereever they need to go?

helmutbuhler commented:
> > One or more required secrets are not set or are empty. R2_ACCESS_KEY_ID: False, R2_SECRET_ACCESS_KEY: False, R2_ENDPOINT_URL: False
> 
> @tintinhamans Can you supply these things whereever they need to go?

They are already in this repository. It's just that they are hidden for pending PRs for security reasons. That's why it's failing currently. Once this PR is merged, all CI Jobs will work. Also all Pending PRs will work from then on, because the data will be in the cache.
If you want to inspect how the run will look, you can look here: https://github.com/helmutbuhler/CnC_Generals_Zero_Hour/pull/1

xezon commented on .github/workflows/check-replays.yml:
```diff
@@ -0,0 +1,236 @@
+          # Generals data files.
+
+          $regPath = "HKCU:\SOFTWARE\Electronic Arts\EA Games\Generals"
+          $installPath = "C:\GameData\Generals\"
```
This script maintains multiple variables with the same strings:

1. `path: C:\GameData`
2. `-oC:\GameData\Generals`
3. `$installPath = "C:\GameData\Generals\"`
4. `$source = "C:\GameData\${{ inputs.game }}"`

I think it would be good to share the names.

So for example

```
$gamePath = "C:\GameData"
$generalsPath = "${{gamePath}}\Generals"
$generalsMDPath = "${{gamePath}}\GeneralsMD"
```

Then it will be easier to tweak locations without breaking things.

xezon commented on .github/workflows/check-replays.yml:
```diff
@@ -0,0 +1,236 @@
+          # Data\Scripts\MultiplayerScripts.scb
+          # Data\Scripts\SkirmishScripts.scb
+
+          Write-Host "Downloading GameData Generals" -ForegroundColor Cyan
```
Maybe improve title as "Downloading Game Data for Generals"

xezon commented on .github/workflows/check-replays.yml:
```diff
@@ -0,0 +1,236 @@
+          New-Item -ItemType Directory -Path $destination -Force | Out-Null
+          Move-Item -Path "$source\*" -Destination $destination -Force
+
+          $source = "${{ inputs.gamedata }}\Maps"
```
The naming of `inputs.gamedata` is ambiguous with GameData, used for the static game data. Maybe name it `inputs.userdata`.

xezon commented on .github/workflows/check-replays.yml:
```diff
@@ -0,0 +1,236 @@
+      gamedata:
+        required: true
+        type: string
+        description: "path to folder with replays and maps"
```
"Path to folder with replays and maps"

xezon reviewed (commented):
Some minor comments

helmutbuhler commented:
Implemented all suggestions

xezon commented on .github/workflows/check-replays.yml:
```diff
@@ -59,6 +59,10 @@ jobs:
           # missing textures, audio and gui files. But it's enough for replay checking.
           # It's also encrypted because it's not allowed to distribute these files.
 
+          $gamePath = "C:\GameData"
```
Can you try move that into the `env:` so that all steps can use it? Then you can remove all the duplicates.

Put the `env:` before the `steps:`

```
    env:
      GAME_PATH: C:\GameData
      GENERALS_PATH ${{ env.GAME_PATH }}\Generals
      GENERALSMD_PATH ${{ env.GAME_PATH }}\GeneralsMD
```

helmutbuhler replied:
I considered that, but it doesn't seem to work. env: works only in steps.

helmutbuhler replied:
But you can try to make it work if you want to. You can push things here and see in my github if it works.

helmutbuhler replied:
I took some codechanges from my ChatGPT, let's see if it works.

xezon replied:
Hopefully. It looks good to the eye :)

helmutbuhler replied:
Works! My ChatGPT from earlier gave some wrong code.

xezon commented:
GitHub says "This branch cannot be rebased due to conflicts". Cannot merge?

helmutbuhler commented:
Just pushed merge of main into this PR. Maybe that helps.

Mauller commented:
This got merged with a bad commit title
